### PR TITLE
Add features and demos for agents and skills

### PR DESCRIPTION
**Pull request by vfm2**
This pull request adds new demo and documentation content to support advanced GitHub Copilot customization, feature exploration, and test engineering best practices. The most important changes are:

---

**Copilot Agent and Skills Documentation**

* Added `.github/agents/blueprint-mode.agent.md`, a comprehensive agent template describing the "Blueprint Mode" for Copilot, including workflow directives, tool usage policy, self-reflection rubric, and communication guidelines. This enables structured, reproducible, and maintainable code generation with strict adherence to project conventions.
* Added `.github/skills/javascript-typescript-jest/SKILL.md`, providing best practices for JavaScript/TypeScript testing using Jest, covering test structure, mocking, async testing, snapshot testing, React component testing, and common matchers.

---

**Demo and Training Content**

* Added `demos/customize-copilot-option-2.md`, a hands-on demo guiding users through advanced Copilot customization, including instruction-driven outputs, agent modes, and skill-based workflows, with practical challenges and evaluation checklists.
* Added `demos/features-demo-option-2.md`, a feature-focused demo that walks users through improving a gallery modal UX using Copilot inline suggestions, chat, and the Copilot App, as well as reviewing changes with AI-powered code review tools.

These additions provide comprehensive resources for leveraging Copilot's advanced features, improving code quality, and standardizing workflows across teams.

### CODE REVIEW

File: .github/hooks/fix-broken-links/link-fix.ps1 [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+#   • With NO file arguments it only lists the broken links - no alternative
+#     lookups and no prompts.
+# Generic anchor text is flagged as an SEO note either way.
+#
+# Pure PowerShell + .NET (Invoke-WebRequest/regex), plus an optional Copilot CLI
+# hand-off for suggestions.
+# Covers: HTML · Markdown · JS/TS · JSON · CSS · SQL · templates (all via URL scan)
+# Trigger: postToolUse
+
+Set-StrictMode -Off
```
Comment (abelberhane) [inline .github/hooks/fix-broken-links/link-fix.ps1]: Should this be? '''Set-StrictMode -Version 3.0'''

From what I can see disabling strict mode silently allows undefined variables and other errors but I am just learning about this now and may be mistaken!

File: .github/hooks/fix-broken-links/link-fix.ps1 [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    elseif ($ch -eq 'r') {
+      if ($alts.Count -gt 0) {
+        Set-UrlReplacement $file $url $alts[0]; $CHANGED[$file] = $true
+        Write-Host "    replaced -> $($alts[0])"; break
+      }
+      Write-Host "    no suggestion available"
+    }
+    elseif ($ch -match '^[1-9]$') {
+      $idx = [int]$ch
+      if ($idx -lt $alts.Count) {
```
Comment (abelberhane) [inline .github/hooks/fix-broken-links/link-fix.ps1:399]: Copilot found an interesting one here that I am trying to validate and wanted to bring up. When user enters 9, the condition is if ($idx -lt $alts.Count). If there are 9 alternatives and user enters 9, this would try to access index 9 of a 0–8 array.

Theres also a suggested Fix: Change to if ($idx -le $alts.Count) or display only valid numbers.

File: .github/hooks/fix-broken-links/link-fix.ps1 [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+function Find-Variation {
+  param([string]$Url)
+  $p = Split-Url $Url
+  $scheme = $p.Scheme; $hostName = $p.Host; $path = $p.Path
+  $cands = [System.Collections.Generic.List[string]]::new()
+  if ($scheme -eq 'http')  { [void]$cands.Add("https://$hostName$path") }
+  if ($scheme -eq 'https') { [void]$cands.Add("http://$hostName$path") }
+  if ($hostName -like 'www.*') { [void]$cands.Add("$scheme`://$($hostName.Substring(4))$path") }
+  else                         { [void]$cands.Add("$scheme`://www.$hostName$path") }
+  if ($path -and $path -notmatch '/$' -and (($path -split '/')[-1]) -notmatch '\.') {
```
Comment (abelberhane) [inline .github/hooks/fix-broken-links/link-fix.ps1:140]: I am not sure if we want to account for this edge case but the condition ($path -split '/')[-1]) -notmatch '\.' tries to avoid adding slash to files, but it fails for URLs like /api/v1 (no dot). This adds a slash when it shouldn't.

File: .github/agents/blueprint-mode.agent.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+Internally validate the solution against engineering best practices before completion. This is a non-negotiable quality gate.
+
+### Rubric (fixed 6 categories, 1–10 integers)
+
+1. Correctness: Does it meet the explicit requirements?
+2. Robustness: Does it handle edge cases and invalid inputs gracefully?
+3. Simplicity: Is the solution free of over-engineering? Is it easy to understand?
+4. Maintainability: Can another developer easily extend or debug this code?
+5. Consistency: Does it adhere to existing project conventions (style, patterns)?
```
Comment (abelberhane) [inline .github/agents/blueprint-mode.agent.md:114]: Mention fixed 6 categories but I see only 5 atm

File: demos/features-demo-option-2.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+For this challenge, we will improve the photo detail modal behavior in [GalleryGrid](../src/components/gallery/GalleryGrid.tsx). The goal is to make the modal feel polished and easier to use.
+
+You have three options for how to implement this challenge. Feel free to get creative and use the option that best suits your workflow. You can also combine options if you like.
+
+**Definition of Done**
+- Go to the gallery page after running the app locally.
+- Select "View Details" on a photo.
+- [] Pressing Escape closes the modal.
+- [] Clicking outside the modal content closes it.
+- [] The page cannot scroll while the modal is open.
```
Comment (abelberhane) [inline demos/features-demo-option-2.md]: Theses arent rendering as I would expect on the page. https://github.com/ps-copilot-sandbox/copilot-intermediate-gallery-repo/blob/vfm2/features-customizes-demos/demos/features-demo-option-2.md

File: demos/features-demo-option-2.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+1. Open [GalleryGrid](../src/components/gallery/GalleryGrid.tsx) and scan where `selectedPhoto` controls the modal open/close state.
+2. Add a short intent comment near the modal logic (for example: `// Close modal on Escape and disable page scroll while model is open`) on line 29 and pause for an inline suggestion before typing more.
+
+> **_NOTE_** If it doesn't appear, try typing a few more words or pressing Enter to trigger it. like `useEffect(() => {` and pause for a suggestion._
+
+3. Find where the modal is rendered and add a short intent comment (for example: add comment `{/* Close modal on backdrop click */}`) in the empty space after `{/* Photo Detail Modal - Placeholder for future implementation */}` and pause for an inline suggestion before typing more.
+
+> **_NOTE_** If it doesn't appear, try typing a few more words or pressing Enter to trigger it. like `onClick={() => {` and pause for a suggestion.
+
+3. Implement one behavior at a time and use **Next Edit Suggestions** after each accepted change to jump to the next likely edit (Escape handling, then backdrop click, then scroll lock).
```
Comment (abelberhane) [inline demos/features-demo-option-2.md]: 3 is repeated twice, should this be 4 and the one after be 5?

File: demos/features-demo-option-2.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+Use inline suggestions to scaffold a few small UX improvements.
+
+**Features to use**
+- Inline suggestions
+- Next edit suggestions
+
+**Tips to complete:**
+
+1. Open [GalleryGrid](../src/components/gallery/GalleryGrid.tsx) and scan where `selectedPhoto` controls the modal open/close state.
+2. Add a short intent comment near the modal logic (for example: `// Close modal on Escape and disable page scroll while model is open`) on line 29 and pause for an inline suggestion before typing more.
```
Comment (abelberhane) [inline demos/features-demo-option-2.md]: is this model is open or modal is open?

File: demos/features-demo-option-2.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+**Features to use**
+- Plan mode
+- Agent mode
+- Model picker
+- Auto model selection
+
+**Tips to complete:**
+
+- Change agent to _Plan mode_ or _Agent mode_
+  - Use which ever mode you're most unfamiliar with to plan or implement the changes.
```
Comment (abelberhane) [inline demos/features-demo-option-2.md]: i believe which ever is whichever but may be mistaken!

File: demos/features-demo-option-2.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+4. Update the branch from `main` to your **current branch**
+5. Look over the options such as `Interactive`, `Plan`, and `Autopilot` to see which one you want to use for this challenge. You can also try out multiple options if you like.
+6. Use the suggested prompt from Option 2 to implement the changes in GalleryGrid.tsx.
+7. View change by selecting `Changes` on the bottom left of the chat
+8. Look over the other options in that panel. Select `Terminal` and run `npm run dev` to view the changes in your browser.
+9. `cmd + click` on the `http://localhost:3000` link to open the app directly in GitHub Copilot App.
+10. Once you are happy with the changes, click on the dropdown of `Create PR` on the top right to select `Create draft PR`.
+11. Click `Create draft PR` and view it on [GitHub](https://github.com/ps-copilot-sandbox/copilot-intermediate-gallery-repo/pulls) OR in the GitHub Copilot App by selecting `PR #` on the top right of the chat.
+
+Feel free to complete that PR to see the change in you current branch.
```
Comment (abelberhane) [inline demos/features-demo-option-2.md:103]: I think this should be your current branch

File: demos/customize-copilot-option-2.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+```markdown
+/javascript-typescript-jest Now expand the suite with edge-case tests:
+- duplicate upload attempts
+- very large file rejection behavior
+- unsupported MIME type handling
+Then provide a "test maintenance checklist" for future UI changes.
+```
+
+6. Evaluate whether output behaved like a reusable skill execution.
```
Comment (abelberhane) [inline demos/customize-copilot-option-2.md]: this list goes from 1, 2, 3, 4, and then 6. Should this be 5?

File: demos/features-demo-option-2.md (hunk 9 above)
Comment (vfm2) [reply demos/features-demo-option-2.md:103]: The app will create a new branch from the current branch your in. For example, I tested in vfm2/demo-part-2 and the app created a new branch (vfm2-improve-modal-ux) off of vfm2/demo-part-2 instead of main when I ran the promp in step 6. 

- [Example here](https://github.com/ps-copilot-sandbox/copilot-intermediate-gallery-repo/pull/65)

File: .github/agents/blueprint-mode.agent.md (hunk 4 above)
Comment (vfm2) [reply .github/agents/blueprint-mode.agent.md:114]: Updated to 5 👍🏻 resolving

File: demos/features-demo-option-2.md (hunk 5 above)
Comment (vfm2) [reply demos/features-demo-option-2.md]: updated 👍🏻 resolving comment

File: demos/features-demo-option-2.md (hunk 7 above)
Comment (vfm2) [reply demos/features-demo-option-2.md]: updated the second 'model' in the comment to modal 👍🏻 resolving comment

File: demos/features-demo-option-2.md (hunk 6 above)
Comment (vfm2) [reply demos/features-demo-option-2.md]: updated 👍🏻 resolving comment

File: demos/features-demo-option-2.md (hunk 8 above)
Comment (vfm2) [reply demos/features-demo-option-2.md]: updated 👍🏻 resolving comment

File: demos/customize-copilot-option-2.md (hunk 10 above)
Comment (vfm2) [reply demos/customize-copilot-option-2.md]: updated 👍🏻 resolving comment

File: demos/features-demo-option-2.md (hunk 9 above)
Comment (vfm2) [reply demos/features-demo-option-2.md:103]: resolving comment

File: .github/hooks/fix-broken-links/link-fix.ps1 (hunk 3 above)
Comment (vfm2) [reply .github/hooks/fix-broken-links/link-fix.ps1:140]: Good catch! That's definitely an edge case with the current heuristic. Since this script is copied directly from awesome-copilot, I'd prefer to keep it in sync with the upstream version rather than introduce local changes. If we think it's worth fixing, I'd be happy to open an issue or PR against the original repo. LMK!

File: .github/hooks/fix-broken-links/link-fix.ps1 (hunk 1 above)
Comment (vfm2) [reply .github/hooks/fix-broken-links/link-fix.ps1]: Since this script is copied directly from awesome-copilot, I'd prefer to keep it in sync with the upstream version rather than introduce local changes. If we think it's worth fixing, I'd be happy to open an issue or PR against the original repo.

File: .github/hooks/fix-broken-links/link-fix.ps1 (hunk 2 above)
Comment (vfm2) [reply .github/hooks/fix-broken-links/link-fix.ps1:399]: Since this script is copied directly from awesome-copilot, I'd prefer to keep it in sync with the upstream version rather than introduce local changes. If we think it's worth fixing, I'd be happy to open an issue or PR against the original repo.

File: .github/hooks/fix-broken-links/link-fix.ps1 (hunk 3 above)
Comment (abelberhane) [reply .github/hooks/fix-broken-links/link-fix.ps1:140]: That makes complete sense! No need to change it!

File: .github/hooks/fix-broken-links/link-fix.ps1 (hunk 2 above)
Comment (abelberhane) [reply .github/hooks/fix-broken-links/link-fix.ps1:399]: That makes sense! I didnt think of that!

**Review by abelberhane — APPROVED** [review]
LGTM!
